### PR TITLE
Issue #1099 Update issue-list.sh to show issues categorically

### LINKED ISSUE #1099: Update issue-list.sh script to print the issues categorically
During the release, we get the list of issues, categorize it into _Bugs_, _Features_ or _Tasks_ and put it as part of release page.

Currently, `issue-list.sh` script displays the issues line wise:
```
$ ./scripts/release/issue-list.sh -r minishift -m 15
- Issue [#661](https://github.com/minishift/minishift/issues/661) (bug) - Hostfolder configuration should not save the password in clear text
- Issue [#667](https://github.com/minishift/minishift/issues/667) (bug) - Allow usage of OpenShift versions lower than Minishfit's baseline
....
- Issue [#696](https://github.com/minishift/minishift/issues/696) (feature) - Need admin user with cluster-admin role
- Issue [#749](https://github.com/minishift/minishift/issues/749) (feature) - Introduce an addon command \echo\
...
- Issue [#647](https://github.com/minishift/minishift/issues/647) (task) - Allow variable substitution to work for local as well as publication site build
- Issue [#649](https://github.com/minishift/minishift/issues/649) (task) - Avoid the use of 'glog.Errorln' in hostfolder related commands
...
```

It is required to display the issues into categories as : 
```
$ ./scripts/release/issue-list.sh -r minishift -m 15
# Bugs
- Issue [#661](https://github.com/minishift/minishift/issues/661) (bug) - Hostfolder configuration should not save the password in clear text
- Issue [#667](https://github.com/minishift/minishift/issues/667) (bug) - Allow usage of OpenShift versions lower than Minishfit's baseline
....

# Features
- Issue [#696](https://github.com/minishift/minishift/issues/696) (feature) - Need admin user with cluster-admin role
- Issue [#749](https://github.com/minishift/minishift/issues/749) (feature) - Introduce an addon command \echo\
...

# Tasks
- Issue [#647](https://github.com/minishift/minishift/issues/647) (task) - Allow variable substitution to work for local as well as publication site build
- Issue [#649](https://github.com/minishift/minishift/issues/649) (task) - Avoid the use of 'glog.Errorln' in hostfolder related commands
...
```

### PR DESCRIPTION
**Pull request by budhrg**
Fix #1099 

Output now: 
```
$ ./scripts/release/issue-list.sh -r minishift -m 17
# Bugs
- Issue [#1004](https://github.com/minishift/minishift/issues/1004) (bug) - Make basic.feature work for upstream and downstream
...
- Issue [#873](https://github.com/minishift/minishift/issues/873) (bug) - Unable to use private insecure registry during start
# Features
- Issue [#1002](https://github.com/minishift/minishift/issues/1002) (feature) - Allow to specify additional default and blocked docker registries
...
- Issue [#938](https://github.com/minishift/minishift/issues/938) (feature) - Interactively ask for username and password during registration
# Tasks
- Issue [#1029](https://github.com/minishift/minishift/issues/1029) (task) - Improve ability to apply CDK customisations
...
- Issue [#919](https://github.com/minishift/minishift/issues/919) (task) - Document SSH usage with Minishift

### CODE REVIEW

**Comment by LalatenduMohanty** [comment]
> jsonstring: In string which is json embeddable (required for easily embed as part of json body during curl command)

@budhrg Do we have any usecase for the jsonstring type of output from this string.

**Comment by budhrg** [comment]
@LalatenduMohanty @hferentschik  See https://github.com/minishift-bot/minishift/blob/master/centos_ci.sh#L246

**Comment by budhrg** [comment]
@LalatenduMohanty 

> However if we can make sure freatures | bug | task can stay in a single column it will be awesome.

Didn't get you by this. The layout for 'human' format is as according to release page notes. The idea is just get the changelog into 'human' format and one should just copy this output and paste to release page notes.

File: scripts/release/issue-list.sh [hunk 1]
```diff
@@ -40,19 +41,41 @@ function milestone_issues()
...
   issue_list=`echo "$issue_list" | tr -d \"`
 
+  # Replace \ which is left over from above command with "(double quote) and suppress warning
+  issue_list=`echo "$issue_list" | tr '\' '"' 2> /dev/null`
+
   # Adjust the issue links
   issue_list=`echo "$issue_list" | sed -e s/api.github.com.repos/github.com/g`
-  echo "$issue_list"
+
+  bugs="# Bugs"
```
Comment (hferentschik) [inline scripts/release/issue-list.sh]: We should change the order. Features first, then bugs and last tasks. Whether one wants to switch the last two is maybe arguable, but features are most important.

File: scripts/release/issue-list.sh [hunk 2]
```diff
@@ -17,9 +17,10 @@
 programname=$0
 function usage()
 {
-    echo "usage: $programname -r repository -m milestone"
+    echo "usage: $programname -r repository -m milestone -f format"
     echo "  -r  repository name"
     echo "  -m  milestone id"
+    echo "  -f  output format ('human' or 'jsonstring')"
```
Comment (hferentschik) [inline scripts/release/issue-list.sh]: Why do we need that? And even if we need it, what 'jsonstring' produces is not JSON. This is about whether new line characters are printed or now. What is the driving force here for this?

**Comment by hferentschik** [comment]
>  id=$(curl -s "https://api.github.com/repos/minishift-bot/minishift/releases/latest" | jq -r '.id')

This is potentially dangerous. You should pass the explicit release id and not rely on latest.

File: scripts/release/issue-list.sh (hunk 1 above)
Comment (budhrg) [inline scripts/release/issue-list.sh]: Ack

File: scripts/release/issue-list.sh (hunk 2 above)
Comment (budhrg) [inline scripts/release/issue-list.sh]: @hferentschik , check here https://github.com/minishift-bot/minishift/blob/master/centos_ci.sh#L248 its usage.

Comment (hferentschik) [inline scripts/release/issue-list.sh]: I don't think so. right now you are spreading the conversion of this output over three places. Here you are dealing with new lines, somewhere else you are doing some URL encoding (which I believe is not necessary if the payload is JSON) and last but not least you are calling a generate_json function. IMO, this script does not have to have a format flag and I think you don't need to url encode, just call issue-list and create your JSON payload. We already have jq installed, it is the swiss army knife of JSON manipulation. What you are after should be as simple as : 

    ./scripts/release/issue-list.sh -r minishift -m 17   | jq -Rn 'inputs|tojson' | jq -s '{body:  add } '

Convert string to JSON using _tojson_ and then wrap this into the payload for the GitHub API call. 

If you want to pack this into a function, you can pass the default output of the script to your _create_json_ function and in there you do the jq conversion and returns JSON payload for the API call.

**Comment by LalatenduMohanty** [comment]
> Didn't get you by this. The layout for 'human' format is as according to release page notes. The idea is just get the changelog into 'human' format and one should just copy this output and paste to release page notes.

I was talking about the cosmetic change about how the output looks  e..g.

```
```
But this is not important , so it is up to you whether you want to change it or not.

File: scripts/release/issue-list.sh (hunk 2 above)
Comment (budhrg) [inline scripts/release/issue-list.sh]: @hferentschik , Tried above. This is how it is currently looking 

https://github.com/minishift-bot/minishift/releases/v1.2.4

Noticed that whole content is wrapped as single string. Also no `\n` character.

Any idea how to manipulate it through `jq`. Honestly, never used `jq` much before this. Just trying out myself.
screenshot from 2017-07-06 16-40-22

**Comment by budhrg** [comment]
@hferentschik , Updated the PR to just categorically print the issues with suggested ordering
```
$ ./scripts/release/issue-list.sh -r minishift -m 17
# Features
- Issue [#1002](https://github.com/minishift/minishift/issues/1002) (feature) - Allow to specify additional default and blocked docker registries
...
- Issue [#938](https://github.com/minishift/minishift/issues/938) (feature) - Interactively ask for username and password during registration
# Bugs
- Issue [#1004](https://github.com/minishift/minishift/issues/1004) (bug) - Make basic.feature work for upstream and downstream
...
- Issue [#873](https://github.com/minishift/minishift/issues/873) (bug) - Unable to use private insecure registry during start
# Tasks
- Issue [#1029](https://github.com/minishift/minishift/issues/1029) (task) - Improve ability to apply CDK customisations
...
- Issue [#919](https://github.com/minishift/minishift/issues/919) (task) - Document SSH usage with Minishift
```

File: scripts/release/issue-list.sh [hunk 3]
```diff
@@ -40,9 +40,22 @@ function milestone_issues()
...
   issue_list=`echo "$issue_list" | tr -d \"`
 
+  # Replace \ which is left over from above command with "(double quote) and suppress warning
+  issue_list=`echo "$issue_list" | tr '\' '"' 2> /dev/null`
+
   # Adjust the issue links
   issue_list=`echo "$issue_list" | sed -e s/api.github.com.repos/github.com/g`
-  echo "$issue_list"
+
+  features="# Features"
```
Comment (hferentschik) [inline scripts/release/issue-list.sh]: +1, I like how you solved this, but sorting into buckets.

Looking good to me. I think we should add some empty lies though as @LalatenduMohanty suggested. Basically empty lines around the headers. Other than that ok.

Are you now able to take this output and create the JSON in the release script? If not, I have some ideas.

Comment (budhrg) [inline scripts/release/issue-list.sh]: > Looking good to me. I think we should add some empty lies though as @LalatenduMohanty suggested. Basically empty lines around the headers. Other than that ok.

It is not required as on release page, it is visible as required. See here https://github.com/minishift-bot/minishift/releases/tag/v1.2.10

**Comment by budhrg** [comment]
@LalatenduMohanty @hferentschik 

Updated to have empty lines around headers: 
```
 $ ./scripts/release/issue-list.sh -r minishift -m 17
# Features

- Issue [#1002](https://github.com/minishift/minishift/issues/1002) (feature) - Allow to specify additional default and blocked docker registries
...
- Issue [#938](https://github.com/minishift/minishift/issues/938) (feature) - Interactively ask for username and password during registration

# Bugs

- Issue [#1004](https://github.com/minishift/minishift/issues/1004) (bug) - Make basic.feature work for upstream and downstream
...
- Issue [#873](https://github.com/minishift/minishift/issues/873) (bug) - Unable to use private insecure registry during start

# Tasks

- Issue [#1029](https://github.com/minishift/minishift/issues/1029) (task) - Improve ability to apply CDK customisations
...
- Issue [#919](https://github.com/minishift/minishift/issues/919) (task) - Document SSH usage with Minishift
```

**Comment by hferentschik** [comment]
@budhrg - here is something you can try: 

    ./scripts/release/issue-list.sh -r minishift -m 17  | | jq -Rn 'inputs + "\n"' | jq -s '{body:  add }'

**Comment by hferentschik** [comment]
Merged, thanks.

**Pull request closed without merging**